v2/apiserver/internal/core/events.go

```diff
@@ -44,15 +45,37 @@ type Event struct {
-	// Labels convey additional event details for the purposes of matching Events
-	// to subscribed projects. For instance, no subscribers to the "GitHub" Source
-	// and the "push" Type are likely to want to hear about push events for ALL
-	// repositories. If the "GitHub" gateway labels events with the name of the
-	// repository from which the event originated (e.g. "repo=github.com/foo/bar")
-	// then subscribers can utilize those same criteria to narrow their
-	// subscription from all push events emitted by the GitHub gateway to just
-	// those having originated from a specific repository.
-	Labels Labels `json:"labels,omitempty" bson:"labels,omitempty"`
+	// Qualifiers provide critical disambiguation of an Event's type. A Project is
+	// considered subscribed to an Event IF AND ONLY IF (in addition to matching
+	// the Event's Source and Type) it matches ALL of the Event's qualifiers
+	// EXACTLY. To demonstrate the usefulness of this, consider any event from a
+	// hypothetical GitHub gateway. If, by design, that gateway does not intend
+	// for any Project to subscribe to ALL Events (i.e. regardless of which
+	// repository they originated from), then that gateway can QUALIFY Events it
+	// emits into Brigade's event bus with repo=<repository name>. Projects
+	// wishing to subscribe to Events from the GitHub gateway MUST include that
+	// Qualifier in their EventSubscription. Note that the Qualifiers field's
+	// "MUST match" subscription semantics differ from the Labels field's "MAY
+	// match" subscription semantics.
+	Qualifiers Qualifiers `json:"qualifiers,omitempty" bson:"qualifiers,omitempty"` // nolint: lll
+	// Labels convey supplementary Event details that Projects may OPTIONALLY use
+	// to narrow EventSubscription criteria. A Project is considered subscribed to
+	// an Event if (in addition to matching the Event's Source, Type, and
+	// Qualifiers) the Event has ALL labels expressed in the Project's
+	// EventSubscription. If the Event has ADDITIONAL labels, not mentioned by the
+	// EventSubscription, these do not preclude a match. To demonstrate the
+	// usefulness of this, consider any event from a hypothetical Slack gateway.
+	// If, by design, that gateway intends for Projects to select between
+	// subscribing to ALL Events or ONLY events originating from a specific
+	// channel, then that gateway can LABEL Events it emits into Brigade's event
+	// bus with channel=<channel name>. Projects wishing to subscribe to ALL
+	// Events from the Slack gateway MAY omit that Label from their
+	// EventSubscription, while Projects wishing to subscribe to only Events
+	// originating from a specific channel MAY include that Label in their
+	// EventSubscription. Note that the Labels field's "MAY match" subscription
+	// semantics differ from the Qualifiers field's "MUST match" subscription
+	// semantics.
+	Labels map[string]string `json:"labels,omitempty" bson:"labels,omitempty"`
```

Comment:
I notice how the `Qualifiers` field here has its type as `Qualifiers` (to utilize its `MarshalBSONValue` method, right?)  Curious as to why we don't have a need for the same for Labels -- it looks like it _used_ to have that same method, etc.?

Reply:
Good observation. It has to do with how each of the fields is queried.

Technically, the new `Qualifiers` field _is_ the old `Labels` field, simply renamed. The _new_ `Labels` field is the _real_ new thing.

When we match on qualifiers, we are looking for an _exact_ match. (This is what we _used_ to be doing with the labels, but @itowlson pointed out, quite correctly, that such behavior really didn't comport with reasonable assumptions people are likely to have about how "labels" work.) In MongoDB, an _exact_ match on a subdocument requires you to be using `bson.D` instead of `bson.M` and that's basically the reason for the `MarshalBSONValue` function on the `Qualifiers` (nee `Labels`) type. (The difference between `bson.D` and `bson.M` is that the order of the keys is canonicalized in a `bson.D`. `bosn.M` doesn't have that feature. Think of it like canonical JSON vs "regular" JSON.)

The match semantics on the _new_ `Labels` field don't involve making an exact match on a subdocment, so this concern of having to use `bson.D` doesn't present itself.

Reply:
Ah, thank you.  That clarifies things!